influxdb3_wal/src/lib.rs

```diff
@@ -424,20 +429,124 @@ pub struct LastCacheDelete {
     pub name: String,
 }
 
-#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+#[derive(Debug, Clone, Eq, PartialEq)]
 pub struct WriteBatch {
     pub database_id: DbId,
     pub database_name: Arc<str>,
-    pub table_chunks: HashMap<Arc<str>, TableChunks>,
+    pub table_chunks: HashMap<(Arc<str>, TableId), TableChunks>,
     pub min_time_ns: i64,
     pub max_time_ns: i64,
 }
 
+impl Serialize for WriteBatch {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut state = serializer.serialize_struct("WriteBatch", 5)?;
+        state.serialize_field("database_id", &self.database_id)?;
+        state.serialize_field("database_name", &self.database_name)?;
+        state.serialize_field("min_time_ns", &self.min_time_ns)?;
+        state.serialize_field("max_time_ns", &self.max_time_ns)?;
+        state.serialize_field(
+            "table_chunks",
+            &self
+                .table_chunks
+                .iter()
+                .map(|((table_name, table_id), value)| {
+                    (format!("{table_name}//{}", table_id.as_u32()), value)
+                })
+                .collect::<HashMap<String, &TableChunks>>(),
+        )?;
+        state.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for WriteBatch {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct WriteBatchVisitor;
+        impl<'de> Visitor<'de> for WriteBatchVisitor {
+            type Value = WriteBatch;
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("struct WriteBatch")
+            }
+            fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
+            where
+                V: MapAccess<'de>,
+            {
+                let mut database_id = None;
+                let mut database_name = None;
+                let mut table_chunks = None;
+                let mut min_time_ns = None;
+                let mut max_time_ns = None;
+                while let Some(key) = map.next_key()? {
+                    match key {
+                        "database_id" => database_id = Some(DbId::from(map.next_value::<u32>()?)),
+                        "database_name" => database_name = Some(map.next_value::<&str>()?.into()),
+                        "min_time_ns" => min_time_ns = Some(map.next_value::<i64>()?),
+                        "max_time_ns" => max_time_ns = Some(map.next_value::<i64>()?),
+                        "table_chunks" => {
+                            table_chunks = Some(
+                                map.next_value::<HashMap<String, TableChunks>>()?
+                                    .into_iter()
+                                    .map(|(key, value)| {
+                                        let mut keys = key.split("//");
```

Comment:
This feels like it's going to be problematic. I'd prefer to have the name and ID be separate fields. I think the better approach would be to have the key be the int table id and then have the name be a member on the value, which could have the table name and the data.

Reply:
This could be done by flipping the hashmap to
```rust
HashMap<TableId, (Arc<str>, TableChunks)>
```
Serde would then `Serialize`/`Deserialize` it gracefully using derive.

Not sure how gracefully that fits in to the broader change set but it would certainly be nice to not need the custom serialization code.

If you need to have both the name and ID in the hash key then you could define a new-type, e.g.,
```rust
#[derive(/* ... */, Serialize, Deserialize)]
struct WriteBatch {
    /* ... */
    table_chunks: HashMap<TableKey, TableChunks>,
}

struct TableKey(Arc<str>, TableId);
```
Then, implement `Serialize`/`Deserialize` on `TableKey`, vs. having to do it for the whole `WriteBatch` type.

Reply:
Okay tried flipping them but that did not work. The main problem here is that serde will not serialize a tuple in a map to JSON. It just will not work, JSON has no tuple type. We could do:

HashMap<Arc<str>, HashMap<TableId, TableChunks>>

but that seems like a lot. The original point was to add the TableId, but it's beginning to look like that's just not possible. @pauldix I think we can either use the name or the Id, but not both. We could punt on adding it to WriteBatch. It's currently in the catalog and when we have actual optimizations we want to do we can bring it in.

Reply:
I was thinking that you want the TableId as the key in the map. Then the value will have the name. We want the map key to be an int anyway because it's way faster to lookup.

Reply:
So the `TableChunks` struct would have table name as part of it. Probably worth renaming TableChunks to be something more appropriate.

Reply:
Okay I can make these changes work then!

Reply:
Hmmm okay so this approach also is fraught with coming head to head with other design choices. TableChunks is meant to just contain those, adding the table name is proving hard to do. For instance we heavily rely on a Default impl and have no easy way to pass in table names, we'd have to do a lookup, but we don't really have access to the catalog. It's becoming increasingly clear that we have a lot of assumptions baked into using table names everywhere and where we have access into looking things up. I think we can't have both really as it's way too invasive. We either switch fully to Ids everywhere and let the names be stored in the catalog or just stick to names. I'm leaning towards the former.

Reply:
We should move to IDs everywhere, but anything that shows up in the WAL should also have the name in it somewhere. That is, an individual WAL file should be self-describing without having to pair it with a catalog.

The other thing about IDs everywhere is that for user facing things, it should be name. For our internal systems like lookup tables, etc, it should be ids.

Reply:
I guess we could drop the name from the WAL part if it proves too difficult. I'm just thinking about future recoverability & debugability. Easier if the name is there.

Reply:
Would a lookup table of name to Id also being in WriteBatch be a good idea?